### ISSUE TITLE
Eager loading recursive relation causes infinite loop

### ISSUE THREAD

**Issue by albertsgrc** [typeorm/typeorm#3663]
**Issue type:**

[ ] question
[X] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[X] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[X] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

**Entity**

```ts
@Entity()
export class Post {
    @PrimaryGeneratedColumn()
    id: number;

    @OneToMany(type => Post, post => post.parentPost, { eager: true })
    public childPosts: Post[];

    @ManyToOne(type => Post, post => post.childPosts, { nullable: true })
    public parentPost: Post;
}
```

Loading any instance of this entity will cause an infinite loop because of the eager option on the recursive relation `childPosts`. 

Note that this will happen even if no post matches the relation, i.e there are no child posts.

**Comment by Convly** [comment]
I'm facing the same issue.
@pleerock, any thought about this?

**Comment by mcousillas6** [comment]
Having the same issue, any news about this?

**Comment by Nosfistis** [comment]
This looks like a tree structure, and the normal repository does not know when to stop joining. Maybe use a [TreeEntity](https://typeorm.io/#/tree-entities) instead?

**Comment by BogdanStanciu** [comment]
Having the same issue, there are any incoming fix ?

**Comment by Sergio-PB** [comment]
Facing the same here

**Comment by schuerg** [comment]
Is there a mechanism in SQL to stop joining with recursion like this if the parent is null?

I am facing the same issue. Still relevant in version 0.2.37

**Comment by imnotjames** [comment]
No, most SQL dialects don't have easy recursive queries and detecting loops like these are difficult.

This should really just be an error when you try to make this kind of query or when you're building the metadata.
